File: day22/task1.erl
-module(task1).
-export([solve/0]).

%% Definitions:

-define(ENEMY_DAMAGE, 9).
-define(EFFECTS, [
  missile,
  drain,
  shield,
  poison,
  recharge
]).
-define(TIMEOUT, 10000).
-record(battle_state, {
  turn :: player | enemy,
  player_hp :: integer(),
  enemy_hp :: integer(),
  mana :: integer(),
  effects :: [{ atom(), integer() }]
}).

%% Type definitions:

-spec accumulator(MinSpent :: integer()) -> none().
-spec effect_cost(Effect :: atom()) -> integer().
-spec effect_duration(Effect :: atom()) -> integer().
-spec available_effects(Mana :: integer(), ExistingEffects :: [atom()]) -> [atom()].
-spec 
  battle (
    AccPID :: pid(), 
    State :: #battle_state{}, 
    Spent :: integer()
  ) -> none().
-spec solve() -> none().

%% Functions:

accumulator(MinSpent) ->
  receive 
    Spent ->
      accumulator(if
        MinSpent > Spent ->
          io:format("Min spent: ~p\n", [Spent]),
          Spent;
        true ->
          MinSpent
      end)
  after 
    ?TIMEOUT ->
      io:format("Accumulator timeouted, min value is ~p\n", [MinSpent])
  end.

effect_cost(Effect) ->
  case Effect of
    missile -> 53;
    drain -> 73;
    shield -> 113;
    poison -> 173;
    recharge -> 229;
    IllegalEffect ->
      EffectName = atom_to_list(IllegalEffect),
      io:format("Unknown effect: ~s\n", [EffectName]),
      infinity
  end.

effect_duration(Effect) ->
  case Effect of
    shield -> 6;
    poison -> 6;
    recharge -> 5;
    IllegalEffect ->
      EffectName = atom_to_list(IllegalEffect),
      io:format("Unknown effect: ~s\n", [EffectName]),
      infinity
  end.

available_effects(Mana, ExistingEffects) ->
  lists:filter(fun (Effect) ->
    Cheap = effect_cost(Effect) =< Mana,
    Enabled = lists:member(Effect, ExistingEffects),
    Cheap and not Enabled
  end, ?EFFECTS).

battle(_, State, _)
  when State#battle_state.player_hp =< 0 ->
  battle_lost;
battle(AccPID, State, Spent)
  when State#battle_state.enemy_hp =< 0 ->
  AccPID ! Spent;
battle(AccPID, State, Spent) ->
  EffectAtoms = lists:map(fun({ Atom, _ }) -> 
    Atom 
  end, State#battle_state.effects),
  AffectedState = lists:foldl(fun(Effect, PrevState) ->
    case Effect of
      poison -> PrevState#battle_state{
        enemy_hp = PrevState#battle_state.player_hp - 6
      };
      recharge -> PrevState#battle_state{
        mana = PrevState#battle_state.mana + 101
      };
      _ -> PrevState
    end
  end, State, EffectAtoms),
  PastState = AffectedState#battle_state{
    effects = lists:filtermap(fun({Atom, Count}) ->
      NewCount = Count - 1,
      if 
        NewCount =< 0 ->
          false;
        true ->
          {true, {Atom, NewCount}}
      end
    end, AffectedState#battle_state.effects)
  },
  if
    PastState#battle_state.enemy_hp =< 0 ->
      AccPID ! Spent;
    PastState#battle_state.turn =:= player ->
      PastEffects = lists:map(fun({ Atom, _ }) ->
        Atom
      end, PastState#battle_state.effects),
      AvailableEffects = available_effects(PastState#battle_state.mana, PastEffects),
      lists:foreach(fun(Effect) ->
        NextState = 
          case Effect of
            missile -> 
              NextEnemyHP = PastState#battle_state.enemy_hp - 4,
              PastState#battle_state{ enemy_hp = NextEnemyHP };
            drain -> 
              NextEnemyHP = PastState#battle_state.enemy_hp - 2,
              NextPlayerHP = PastState#battle_state.player_hp + 2,
              PastState#battle_state{ 
                enemy_hp = NextEnemyHP, 
                player_hp = NextPlayerHP 
              };
            LongEffect ->
              NextEffects = lists:append(
                PastState#battle_state.effects,
                [{LongEffect, effect_duration(LongEffect)}]
              ),
              PastState#battle_state{ effects = NextEffects }
          end,
        spawn(fun() ->
          battle(AccPID, NextState#battle_state{
            turn = enemy
          }, Spent + effect_cost(Effect))
        end)
      end, AvailableEffects);
    true ->
      NextPlayerHP = 
        case lists:member(shield, EffectAtoms) of
          true ->
            Damage = max(1, ?ENEMY_DAMAGE - 7),
            PastState#battle_state.player_hp - Damage;
          _ ->
            PastState#battle_state.player_hp - ?ENEMY_DAMAGE
        end,
      NextState = PastState#battle_state{
        turn = player,
        player_hp = NextPlayerHP
      },
      spawn(fun() ->
        battle(AccPID, NextState, Spent)
      end)
  end,
  finished.

solve() ->
  AccPID = spawn(fun() ->
    accumulator(infinity)
  end),
  InitialState = 
    #battle_state{
      turn = player,
      player_hp = 50,
      enemy_hp = 51,
      mana = 500,
      effects = []
    },
  battle(AccPID, InitialState, 0).

File: day22/task1.go
package main

import (
	"errors"
	"fmt"
	"sync"
)

const (
  PoisonDamage = 3
  RechargeMana = 101
  ShieldProtection = 7
  MagicMissileDamage = 4
  DrainDamage = 2
  DrainHeal = 2
)

type Turn bool
const (
  Player Turn = true
  Enemy Turn = false
)

type Effect int
const (
  Shield Effect = iota
  Poison
  Recharge
  MagicMissile
  Drain
)

func Effects() []Effect {
  return []Effect{
    Shield,
    Poison,
    Recharge,
    MagicMissile,
    Drain,
  }
}

func (e Effect) Price() (int, error) {
  switch e {
  case MagicMissile:
    return 53, nil
  case Drain:
    return 73, nil
  case Shield:
    return 113, nil
  case Poison:
    return 173, nil
  case Recharge:
    return 229, nil
  default:
    return 0, errors.New("Unknown effect")
  }
}

func (e Effect) Duration() (int, error) {
  switch e {
  case Recharge:
    return 5, nil
  case Shield:
    return 6, nil
  case Poison:
    return 6, nil
  default:
    return 0, errors.New("This effect doesn't have duration")
  }
}

type BattleState struct {
  Turn Turn
  Effects [3]int
  PlayerHP int
  EnemyHP int
  EnemyDamage int
  Mana int
  SpentMana int
}

func (bs BattleState) Next() []BattleState {
  if bs.PlayerHP <= 0 || bs.EnemyHP <= 0 {
    return []BattleState{}
  }
  if bs.Effects[Poison] > 0 {
    bs.EnemyHP -= PoisonDamage
    bs.Effects[Poison]--
    if bs.EnemyHP <= 0 {
      return []BattleState{bs}
    }
  }
  if bs.Effects[Recharge] > 0 {
    bs.Mana += RechargeMana
    bs.Effects[Recharge]--
  }
  if bs.Turn == Enemy {
    var damage int
    if bs.Effects[Shield] > 0 {
      damage = max(1, bs.EnemyDamage - ShieldProtection)
      bs.Effects[Shield]--
    } else {
      damage = bs.EnemyDamage
    }
    bs.PlayerHP -= damage
    bs.Turn = Player
    return []BattleState{bs}
  }
  next := []BattleState{}
  for _, effect := range Effects() {
    price, err := effect.Price()
    if err != nil {
      fmt.Println(err)
      continue
    }
    if price > bs.Mana {
      continue
    }
    state := bs
    switch effect {
    case MagicMissile:
      state.EnemyHP -= MagicMissileDamage
    case Drain:
      state.EnemyHP -= DrainDamage
      state.PlayerHP += DrainHeal
    case Shield, Poison, Recharge:
      duration, err := effect.Duration()
      if err != nil {
        fmt.Println(err)
        continue
      }
      state.Effects[effect] = duration
    }
    state.Mana -= price
    state.SpentMana += price
    state.Turn = Enemy
    next = append(next, state)
  }
  return next
}

func main() {
  configuration := BattleState{
    Turn: Player,
    Effects: [3]int{0, 0, 0},
    PlayerHP: 50,
    EnemyHP: 51,
    EnemyDamage: 9,
    Mana: 500,
    SpentMana: 0,
  }

  manaSpent := make(chan int)
  done := make(chan struct{})
  go func() {
    minValue := int(^uint(0) >> 1)
    for {
      select {
      case value := <- manaSpent:
        minValue = min(minValue, value)
        fmt.Printf("Received: %d, min: %d\n", value, minValue)
      case <-done:
        fmt.Printf("Finished with min mana: %d\n", minValue)
      }
    }
  }()

  states := []BattleState{configuration}
  for len(states) > 0 {
    nextStates := []BattleState{}
    var wg sync.WaitGroup
    wg.Add(len(states))
    results := make(chan BattleState)
    for _, state := range states {
      go func(state BattleState) {
        defer wg.Done()
        if state.EnemyHP <= 0 {
          manaSpent <- state.SpentMana
        }
        for _, nextState := range state.Next() {
          results <- nextState
        }
      } (state)
    }
    go func() {
      defer close(results)
      wg.Wait()
    }()
    for nextState := range results {
      nextStates = append(nextStates, nextState)
    }
    states = nextStates
  }
  
  done <- struct{}{}
}
